Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -198,6 +221,267 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+
+                scripts.append(format.toString());
+                continue;
+            }
+
+            format.append("importScripts(\""_s, scriptPath, "\");"_s);
+
+            scripts.append(format.toString());
+            continue;
+        }
+
+        format.append("<script"_s);
+        if (usesModules)
+            format.append(" type=\"module\" "_s);
+        format.append("src=\""_s, scriptPath, "\"></script>"_s);
+
+        scripts.append(format.toString());
+    }
+
+    StringBuilder generatedBackgroundContent;
+    if (!isServiceWorker)
+        generatedBackgroundContent.append("<!DOCTYPE html>\n<body>\n"_s);
+
+    for (auto scriptPath : scripts)
+        generatedBackgroundContent.append(scriptPath, "\n"_s);
+
+    if (!isServiceWorker)
+        generatedBackgroundContent.append("\n</body>"_s);
+
+    m_generatedBackgroundContent = generatedBackgroundContent.toString();
+    return m_generatedBackgroundContent;
+}
+
+void WebExtension::populateBackgroundPropertiesIfNeeded()
+{
+    if (m_parsedManifestBackgroundProperties)
+        return;
+
+    m_parsedManifestBackgroundProperties = true;
+
+    RefPtr manifestObject = this->manifestObject();
+    if (!manifestObject)
+        return;
+
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/manifest.json/background
+
+    RefPtr backgroundManifestObject = manifestObject->getObject(backgroundManifestKey);
+    if (!backgroundManifestObject || !backgroundManifestObject->size()) {
+        if (manifestObject->getValue(backgroundManifestKey))
+            recordError(createError(Error::InvalidBackgroundContent));
+        return;
+    }
+
+    m_backgroundPagePath = backgroundManifestObject->getString(backgroundPageManifestKey);
+    m_backgroundServiceWorkerPath = backgroundManifestObject->getString(backgroundServiceWorkerManifestKey);
+    m_backgroundContentUsesModules = (backgroundManifestObject->getString(backgroundPageTypeKey) == backgroundPageTypeModuleValue);
+
+    if (RefPtr backgroundScriptPaths = backgroundManifestObject->getArray(backgroundScriptsManifestKey)) {
+        backgroundScriptPaths = filterObjects(*backgroundScriptPaths, [](auto& value) {
+            return !value.asString().isEmpty();
+        });
+
+        m_backgroundScriptPaths = makeStringVector(*backgroundScriptPaths);
+    }
+
+    HashSet<String> supportedEnvironments = { backgroundDocumentManifestKey, backgroundServiceWorkerManifestKey };
+
+    Vector<String> preferredEnvironments;
+    if (auto environment = backgroundManifestObject->getString(backgroundPreferredEnvironmentManifestKey); !environment.isEmpty()) {
+        if (supportedEnvironments.contains(environment))
+            preferredEnvironments.append(environment);
+    } else if (RefPtr environments = backgroundManifestObject->getArray(backgroundPreferredEnvironmentManifestKey)) {
+        Ref filteredEnvironments = filterObjects(*environments, [supportedEnvironments](auto& value) {
```

Comment:
This here isn't working properly it seems, the BackgroundPreferredEnvironmentParsing test crashes here due to an assertion failure inside the HashSet. I'm not sure if it's not capturing the right values or what, but something with the HashSet in the lambda isn't working.

Reply:
Switching supportedEnvironments back to a Vector seems to work better here.